validator/validator.go

```diff
@@ -99,62 +99,114 @@ func (v *Validator) ValidateToken(ctx context.Context, tokenString string) (inte
 		return nil, fmt.Errorf("could not parse the token: %w", err)
 	}
 
-	if string(v.signatureAlgorithm) != token.Headers[0].Algorithm {
-		return nil, fmt.Errorf(
-			"expected %q signing algorithm but token specified %q",
-			v.signatureAlgorithm,
-			token.Headers[0].Algorithm,
-		)
+	if err = validateSigningMethod(string(v.signatureAlgorithm), token.Headers[0].Algorithm); err != nil {
+		return nil, fmt.Errorf("signing method is invalid: %w", err)
 	}
 
-	key, err := v.keyFunc(ctx)
+	registeredClaims, customClaims, err := v.deserializeClaims(ctx, token)
 	if err != nil {
-		return nil, fmt.Errorf("error getting the keys from the key func: %w", err)
-	}
-
-	claimDest := []interface{}{&jwt.Claims{}}
-	if v.customClaims != nil && v.customClaims() != nil {
-		claimDest = append(claimDest, v.customClaims())
+		return nil, fmt.Errorf("failed to deserialize token claims: %w", err)
 	}
 
-	if err = token.Claims(key, claimDest...); err != nil {
-		return nil, fmt.Errorf("could not get token claims: %w", err)
+	if err = validateClaimsWithLeeway(registeredClaims, v.expectedClaims, v.allowedClockSkew); err != nil {
+		return nil, fmt.Errorf("expected claims not validated: %w", err)
 	}
 
-	registeredClaims := *claimDest[0].(*jwt.Claims)
-	expectedClaims := v.expectedClaims
-	expectedClaims.Time = time.Now()
-	if err = registeredClaims.ValidateWithLeeway(expectedClaims, v.allowedClockSkew); err != nil {
-		return nil, fmt.Errorf("expected claims not validated: %w", err)
+	if customClaims != nil {
+		if err = customClaims.Validate(ctx); err != nil {
+			return nil, fmt.Errorf("custom claims not validated: %w", err)
+		}
 	}
 
 	validatedClaims := &ValidatedClaims{
 		RegisteredClaims: RegisteredClaims{
-			Issuer:   registeredClaims.Issuer,
-			Subject:  registeredClaims.Subject,
-			Audience: registeredClaims.Audience,
-			ID:       registeredClaims.ID,
+			Issuer:    registeredClaims.Issuer,
+			Subject:   registeredClaims.Subject,
+			Audience:  registeredClaims.Audience,
+			ID:        registeredClaims.ID,
+			Expiry:    numericDateToUnixTime(registeredClaims.Expiry),
+			NotBefore: numericDateToUnixTime(registeredClaims.NotBefore),
+			IssuedAt:  numericDateToUnixTime(registeredClaims.IssuedAt),
 		},
+		CustomClaims: customClaims,
 	}
 
-	if registeredClaims.Expiry != nil {
-		validatedClaims.RegisteredClaims.Expiry = registeredClaims.Expiry.Time().Unix()
+	return validatedClaims, nil
+}
+
+func validateClaimsWithLeeway(actualClaims jwt.Claims, expected jwt.Expected, leeway time.Duration) error {
```

Comment:
This replaces https://github.com/square/go-jose/blob/c9ac459e06bc9a5ab7efd5d3b3f421d6de27401f/jwt/validation.go#L76.

Reply:
Should this also include the issuer and jti validation?

Reply:
Although that original func does include them, we were never actually using them for the validation. We're only setting as expected the audience, issuer and the time based claims. 

Re: 
- https://github.com/auth0/go-jwt-middleware/blob/master/validator/validator.go#L82-L85

Reply:
Note that in other SDKs we do validate the sub claim: https://github.com/auth0/Auth0.swift/blob/master/Auth0/IDTokenValidator.swift#L56

Reply:
Maybe we can add that validation?

Reply:
(if applicable, of course).

Reply:
Unfortunately we can't introduce that right now without a breaking change. The `expectedClaims` are a private internal property inside the validator and we set it with the passed in issuer and audience when constructing the validator. We'll have to change this func https://github.com/auth0/go-jwt-middleware/blob/master/validator/validator.go#L59 to allow for the subject as well. I'll note this down for a potential upcoming v3 perhaps, wdyt?